Implementation of abstraction for sending data through MPI

albestro opened the pull request:
Close #62

- Implement `Buffer` concept, a pointer to an underlying buffer with information about the layout, how it can be seen in terms of "one or more blocks with a fixed number of elements (optionally strided)".
- Implement `message` object that wraps data that can be sent with MPI. It uses the Buffer concept to access information and data of the buffer to be sent.
- Implement `type_handler` object that creates a custom MPI Datatype (e.g. MPI_Type_vector) based on the data given. It owns the MPI resource, so once it is deleted it calls MPI_Free_type.
- Extend `mpi_datatype` implementation with more types (even `const` one are mapped).

The broadcast API and tests have been adapted.

albestro commented:
It is not the final API, it is just to give an idea of what it could it be, for discussing and commenting it.

albestro commented on include/dlaf/util_tile.h:
```diff
@@ -0,0 +1,26 @@
+
+/// @brief Create a common::Buffer from a Tile
+template <class T, Device device>
+auto create_buffer(const dlaf::Tile<T, device>& tile) {
```
@rasolca 
Do you have something against embedding this function into `tile.h`?

Since `Buffer` and `Message` concepts are separated, it wouldn't add any dependency to MPI.

Instead, having it in a separate file like this forces users to remember to include it every time they need to send a `Tile`, which IMHO is not straightforward.

albestro commented on include/dlaf/common/buffer.h:
```diff
@@ -0,0 +1,149 @@
+
+/// @brief Helper class for creatig a buffer from a bounded array
+template <class T, std::size_t N>
+struct Buffer<T[N], std::enable_if_t<std::is_array<T[N]>::value>> : Buffer<std::decay_t<T[N]>> {
```
Yes, you're right. It was there due to an additional specialization I was developing, but it is not used anymore.

albestro commented on test/unit/communication/test_broadcast_tile.cpp:
```diff
@@ -0,0 +1,55 @@
+
+  for (int j = 0; j < tile.size().cols(); ++j)
+    for (int i = 0; i < tile.size().rows(); ++i)
+      EXPECT_EQ(message_values({i, j}), *tile.ptr({i, j}));
```
Good hint! 😉

Please, give a look at the code, since I added `CHECK_TILE_NE` and to do this I had to slightly change the code behind.

rasolca commented on test/unit/communication/test_broadcast_tile.cpp:
```diff
@@ -12,44 +12,42 @@
-    for (int j = 0; j < tile.size().cols(); ++j)
-      for (int i = 0; i < tile.size().rows(); ++i)
-        EXPECT_NE(message_values({i, j}), *tile.ptr({i, j}));
+    CHECK_TILE_NE(message_values, tile);
```
Avoid checks on unitialized memory. It is not guaranteed that the memory of the tile is different from the message value.
Prefer the approach which sets the elements (e.g. set to NaN or also 0 since all your elements (real and imaginary parts for complex) are >0).

albestro commented on test/unit/test_tile.cpp:
```diff
@@ -334,3 +334,25 @@ TYPED_TEST(TileTest, PromiseToFutureConst) {
   CHECK_TILE_PTR(ptr, tile2);
 }
+
+// Buffer
```
After https://github.com/eth-cscs/DLA-Future/pull/70#discussion_r341028776 `Tile` now includes the `Buffer`.

albestro commented:
As discussed, I removed from this PR the async broadcast code (there will be another PR about it) and I have also reverted the wrong code change about `CHECK_TILE_NE` and replaced it with a `set` function that accepts a single element instead of a functor.

albestro commented:
Close #58 

I closed PR #60 since everything is in this one now.

rasolca commented on test/include/dlaf_test/util_tile.h:
```diff
@@ -38,6 +40,21 @@ void set(const Tile<T, Device::CPU>& tile, Func el) {
+/// @brief Sets all the elements of the tile with given value.
+///
+/// @pre el argument is an index of type const TileElementIndex&.
+/// @pre el return type should be T.
```
This doc is a copy and paste from the previous. Please update it.

rasolca commented on test/include/dlaf_test/util_tile.h:
```diff
@@ -38,6 +40,21 @@ void set(const Tile<T, Device::CPU>& tile, Func el) {
+/// @pre el return type should be T.
+template <class T, class U, enable_if_convertible_t<U, T, int> = 0>
+void set(const Tile<T, Device::CPU>& tile, U el) {
+  // TODO evaluate to use the other one with a lambda
```
IMHO both approaches are OK.
Fix this TODO before merging.
